Additions/improvements for doc examples.

frewsxcv opened the pull request.

carols10cents commented:
r? @GuillaumeGomez

ollie27 commented on src/libstd/ffi/c_str.rs:
````diff
@@ -844,6 +950,21 @@ impl CStr {
+    /// ```
+    ///
+    /// Had `c_str` contained invalid unicode, the `to_string_lossy` call might
+    /// have returned `Cow::Borrowed("fo�")`.
````
How about adding an example for the invalid Unicode case? Either way is definitely wouldn't return `Cow::Borrowed` it would return `Cow::Owned`.

frewsxcv replied:
I figured it'd be undefined behavior to create a string w/ invalid unicode, as [your comment here](https://github.com/rust-lang/rust/pull/38839#discussion_r94711507). 

> Either way is definitely wouldn't return Cow::Borrowed it would return Cow::Owned.

Yep, good catch, I'll change that

ollie27 replied:
> I figured it'd be undefined behavior to create a string w/ invalid unicode

`String`s and `str`s have to contain valid UTF-8 but `CString`s and `Cstr`s are just sequences of any none null bytes. The following example would be fine:

```rust
use std::borrow::Cow;
use std::ffi::CStr;

let c_str = CStr::from_bytes_with_nul(b"foo\xc3\0").unwrap();
assert_eq!(c_str.to_string_lossy(), "foo�");
```

It would also be more correct to say "(in)valid UTF-8" rather than "(in)valid unicode" in this case.

QuietMisdreavus replied:
It might be worth borrowing from the example in `String::from_utf8_lossy`, which uses the input string `b"Hello \xF0\x90\x80World"` to demonstrate the replacement.

QuietMisdreavus commented:
Looks good, other than the discussion surrounding `to_string_lossy`. If you can update those examples (on both `CStr` and `OsStr`) then I'm okay with everything else.

frewsxcv commented:
> Looks good, other than the discussion surrounding `to_string_lossy`. If you can update those examples (on both `CStr` and `OsStr`) then I'm okay with everything else.

Fixed the failing doc test and updated the `CStr::to_string_lossy` examples. I ditched my changes to `OsStr`. I don't think there's a way to create an `OsStr` from bytes (in a cross platform manner) like we can for `CStr` so I don't know how to demonstrate an `OsStr` containing invalid UTF8.

kennytm commented on src/libstd/ffi/c_str.rs:
````diff
@@ -271,6 +287,27 @@ impl CString {
+    /// Create a `CString`, pass ownership to an `extern` function (via raw pointer), then retake
+    /// ownership with `from_raw`:
+    ///
+    /// ```ignore
````
Could you change this to  ` ```no_run `? Also the `some_external_function(raw)` on line 307 should be `some_extern_function(raw)`.

frewsxcv replied:
I think I'll get linker errors upon building, right?

frewsxcv replied:
Latest force push addresses the typo, good catch

kennytm replied:
@frewsxcv Currently `no_run` (and `compile_fail` as well) will [stop after analysis](https://github.com/rust-lang/rust/blob/master/src/librustdoc/test.rs#L248-L250). It won't even run codegen.

frewsxcv commented:
Changed `ignore` to `no_run` in the latest force push

bors commented:
:umbrella: The latest upstream changes (presumably #42716) made this pull request unmergeable. Please resolve the merge conflicts.

frewsxcv commented:
Rebased, addressed merge conflicts

QuietMisdreavus commented:
@bors r+

bors commented:
:pushpin: Commit 58bbe1d has been approved by `QuietMisdreavus`

<!-- @bors r=QuietMisdreavus 58bbe1d68cddacfd8dca792adb40a6dd04e6d319 -->

frewsxcv commented:
@bors rollup
